CBBD-385: Implemented the TransformerUtils mapEobType method and adde…

glovercloud opened the pull request:
…d corresponding test updates and documentation.

dho937 commented on bluebutton-server-app/src/test/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/CarrierClaimTransformerTest.java:
```diff
@@ -246,5 +245,16 @@ static void assertMatches(CarrierClaim claim, ExplanationOfBenefit eob) throws F
+				String.valueOf(claim.getNearLineRecordIdCode()), eob.getType());
+		TransformerTestUtils.assertHasCoding(TransformerConstants.CODING_NCH_CLAIM_TYPE,
+				claim.getClaimTypeCode(), eob.getType());
 	}
```
This is very minor but you might be able to make a common assertEobType method in TransformerTestUtils like you did for the mapping to do all of these assertions.  Just pass in the fields you need to the assertEobType method.  Again, it looks fine the way it is so no big deal.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/CarrierClaimTransformer.java:
```diff
@@ -67,10 +67,11 @@ private static ExplanationOfBenefit transformClaim(CarrierClaim claimGroup) {
 
 		eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
```
This should also be handled by your new method and removed here.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/DMEClaimTransformer.java:
```diff
@@ -68,10 +68,11 @@ private static ExplanationOfBenefit transformClaim(DMEClaim claimGroup) {
 
 		eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
```
This should also be handled by your new method and removed here.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/HHAClaimTransformer.java:
```diff
@@ -60,10 +60,11 @@ private static ExplanationOfBenefit transformClaim(HHAClaim claimGroup) {
 
 		eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
```
This should also be handled by your new method and removed here.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/HospiceClaimTransformer.java:
```diff
@@ -58,14 +58,13 @@ private static ExplanationOfBenefit transformClaim(HospiceClaim claimGroup) {
 
-		eob.setType(TransformerUtils.createCodeableConcept(
-				TransformerConstants.CODING_CCW_CLAIM_TYPE,
+		eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
```
This should also be handled by your new method and removed here.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/OutpatientClaimTransformer.java:
```diff
@@ -57,10 +57,11 @@ private static ExplanationOfBenefit transformClaim(OutpatientClaim claimGroup) {
 
 		eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
```
This should also be handled by your new method and removed here.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/PartDEventTransformer.java:
```diff
@@ -53,10 +54,8 @@ private static ExplanationOfBenefit transformClaim(PartDEvent claimGroup) {
 		 * Note: Part D events are the one CCW "claim" type that don't have
```
I'd remove/move this comment -- it makes more sense to include in your new method.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/SNFClaimTransformer.java:
```diff
@@ -60,10 +60,11 @@ private static ExplanationOfBenefit transformClaim(SNFClaim claimGroup) {
 
 		eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
```
This should also be handled by your new method and removed here.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/TransformerConstants.java:
```diff
@@ -298,6 +299,9 @@
 
 	public static final String CODING_CCW_MCO_PAID = "https://www.ccwdata.org/cs/groups/public/documents/datadictionary/mcopdsw.txt";
 
+	// FIXME: there really isn't an NCH section, should this be moved somewhere else within this file?
```
Make sure that this constant appears in alphabetical order, per the class JavaDoc.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/TransformerUtils.java:
```diff
@@ -87,7 +88,7 @@ static CareTeamComponent addCareTeamPractitioner(ExplanationOfBenefit eob, ItemC
 	
 		// Link the EOB.item to the care team entry (if it isn't already).
-		if (!eobItem.getCareTeamLinkId().contains(careTeamEntry.getSequence())) {
+		if (!eobItem.getCareTeamLinkId().contains(new PositiveIntType(careTeamEntry.getSequence()))) {
```
I don't think this change belongs in this PR? No idea if it's a good idea or not, but it's out of place.

glovercloud replied:
Eclipse Oxygen was flagging it as a warning so I just fixed it.

glovercloud replied:
Will revert and added a card [CBBF-124](https://issues.hhsdevcloud.us/browse/CBBF-124)

karlmdavis commented on dev/data-model.md:
```diff
@@ -84,6 +84,8 @@ The following [Coding](http://www.hl7.org/implement/standards/fhir/datatypes.htm
 The following [Identifier](http://hl7.org/fhir/STU3/datatypes.html#identifier) systems can be found in API responses:
 
+* `https://bluebutton.cms.gov/developer/docs/reference/some-thing` 
+	* temporary placeholder for claim type link
```
:+1: Thanks for adding this!

glovercloud commented:
I wasn’t sure what to put in the api-changelog as the api did not change just the implementation.  You just want the issue linked?

On Jan 12, 2018, 2:26 PM -0700, Karl M. Davis <notifications@github.com>, wrote:
> @karlmdavis requested changes on this pull request.
> Some minor changes needed. But please also update the dev/api-changelog.md file!
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/CarrierClaimTransformer.java:
> > @@ -67,10 +67,11 @@ private static ExplanationOfBenefit transformClaim(CarrierClaim claimGroup) {
>
>                eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
> This should also be handled by your new method and removed here.
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/DMEClaimTransformer.java:
> > @@ -68,10 +68,11 @@ private static ExplanationOfBenefit transformClaim(DMEClaim claimGroup) {
>
>                eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
> This should also be handled by your new method and removed here.
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/HHAClaimTransformer.java:
> > @@ -60,10 +60,11 @@ private static ExplanationOfBenefit transformClaim(HHAClaim claimGroup) {
>
>                eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
> This should also be handled by your new method and removed here.
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/HospiceClaimTransformer.java:
> > @@ -58,14 +58,13 @@ private static ExplanationOfBenefit transformClaim(HospiceClaim claimGroup) {
>                eob.addIdentifier().setSystem(TransformerConstants.CODING_CCW_CLAIM_GROUP_ID)
>                                .setValue(claimGroup.getClaimGroupId().toPlainString());
>
> -               eob.setType(TransformerUtils.createCodeableConcept(
> -                               TransformerConstants.CODING_CCW_CLAIM_TYPE,
> +               eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
> This should also be handled by your new method and removed here.
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/OutpatientClaimTransformer.java:
> > @@ -57,10 +57,11 @@ private static ExplanationOfBenefit transformClaim(OutpatientClaim claimGroup) {
>
>                eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
> This should also be handled by your new method and removed here.
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/PartDEventTransformer.java:
> > @@ -53,10 +54,8 @@ private static ExplanationOfBenefit transformClaim(PartDEvent claimGroup) {
>                 * Note: Part D events are the one CCW "claim" type that don't have
> I'd remove/move this comment -- it makes more sense to include in your new method.
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/SNFClaimTransformer.java:
> > @@ -60,10 +60,11 @@ private static ExplanationOfBenefit transformClaim(SNFClaim claimGroup) {
>
>                eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
> This should also be handled by your new method and removed here.
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/TransformerConstants.java:
> > @@ -298,6 +299,9 @@
>
>        public static final String CODING_CCW_MCO_PAID = "https://www.ccwdata.org/cs/groups/public/documents/datadictionary/mcopdsw.txt";
>
> +       // FIXME: there really isn't an NCH section, should this be moved somewhere else within this file?
> Make sure that this constant appears in alphabetical order, per the class JavaDoc.
> In bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/TransformerUtils.java:
> > @@ -87,7 +88,7 @@ static CareTeamComponent addCareTeamPractitioner(ExplanationOfBenefit eob, ItemC
>                }
>
>                // Link the EOB.item to the care team entry (if it isn't already).
> -               if (!eobItem.getCareTeamLinkId().contains(careTeamEntry.getSequence())) {
> +               if (!eobItem.getCareTeamLinkId().contains(new PositiveIntType(careTeamEntry.getSequence()))) {
> I don't think this change belongs in this PR? No idea if it's a good idea or not, but it's out of place.
> In dev/data-model.md:
> > @@ -84,6 +84,8 @@ The following [Coding](http://www.hl7.org/implement/standards/fhir/datatypes.htm
>
> The following [Identifier](http://hl7.org/fhir/STU3/datatypes.html#identifier) systems can be found in API responses:
>
> +* `https://bluebutton.cms.gov/developer/docs/reference/some-thing`
> +       * temporary placeholder for claim type link
> 👍 Thanks for adding this!
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

karlmdavis commented on dev/api-changelog.md:
```diff
@@ -1,5 +1,9 @@
 
+## CBBD-385
+
+* Created the TransformerUtils method mapEobType to map certain CCW codings into a corresponding FHIR coding where appropriate.
```
Can you please document this more from the perspective of a developer using the API? What response fields were added/changed/removed?

karlmdavis commented on dev/api-changelog.md:
```diff
@@ -1,6 +1,23 @@
+        * Please note that this `Coding` had previously been mapped to an extension.
+    4. TODO: near line record ID
+
+## CBBF-92 (Sprint 42, 2018-01)
```
Went through and partially reworked this, with a couple `TODO`s left for you to fill in. Please note the focus here on how our third-party developers will view the changes.

Once we have solid documentation of our API's responses, the API changelog won't need to be quite so explicit. But -- we don't yet, so it's best to provide a decent amount of detail here for now.

glovercloud replied:
Why don't you just fill in the TODOs because it is not clear to me how you want them described.  Also your ClaimType link was broken syntactically.  This way we can get this completed and merged.

glovercloud replied:
Thanks Karl, looks great!  I don't think I would have been able to produce the detail you did here but I now know what you are looking for in the future.  I will merge to master.

karlmdavis commented on bluebutton-server-app/src/main/java/gov/hhs/cms/bluebutton/server/app/stu3/providers/TransformerUtils.java:
```diff
@@ -516,4 +516,72 @@ static void validatePeriodDates(Optional<LocalDate> dateFrom, Optional<LocalDate
+			eob.setType(TransformerUtils.createCodeableConcept(TransformerConstants.CODING_CCW_CLAIM_TYPE,
+					ccwClaimTypeCode.get()));
+			eob.getType().addCoding().setSystem(TransformerConstants.CODING_NCH_CLAIM_TYPE).setCode(
+					ccwClaimTypeCode.get());
```
The possible values for this field are (like most coded fields) going to be listed in the [Medicare Fee-For-Service Claims codebook](https://www.ccwdata.org/documents/10280/19022436/codebook-ffs-claims.pdf) codebook (found on the [CCW Data Dictionaries](https://www.ccwdata.org/web/guest/data-dictionaries) page). In the API changelog, just point devs to the PDF and mention the field name.

We are shortly going to be replacing the PDF with linkable pages, that you'd be able to reference instead... but that's not ready yet.

karlmdavis commented:
@glovercloud Can you please review my last 3 commits, then merge if approved?

@dho937 Please take a look at the API changelog here, just as an FYI. This is the level of detail we want to provide for these kinds of changes at the moment. (Once we have better docs in place for third-party developers, we'll be able to mostly just reference them when things are added, but breaking changes will always require a lot of detail here, I think.)
